### PR TITLE
[WIP] Added implementation of SGHMC and example.

### PR DESCRIPTION
**Pull request by dwadden**
I'm working with @nfoti and @yianma. We have some questions on a few small parts of the code. See comments in changed files.

### CODE REVIEW

File: edward/inferences/sghmc.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    return super(SGHMC, self).initialize(*args, **kwargs)
+
+  def build_update(self):
+    """
+    Simulate Hamiltonian dynamics with friction using a discretized
+    integrator. Its discretization error goes to zero as the learning rate
+    decreases.
+    Implements the update equations from (15) of Chen et al., 2014.
+    """
+    # TODO: Would it be more memory-efficient to use tf.control_dependencies,
```
Comment (dwadden) [inline edward/inferences/sghmc.py]: Does making aliases make copies of the tensors? If so, would it save memory to use `tf.control_dependencies` and mutate the objects directly?

File: edward/inferences/sghmc.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    # rather than repeatedly re-assign attributes stored on self.v?
+    old_sample = {z: tf.gather(qz.params, tf.maximum(self.t - 1, 0))
+                  for z, qz in six.iteritems(self.latent_vars)}
+    old_v_sample = {z: v for z, v in six.iteritems(self.v)}
+
+    # Simulate Hamiltonian dynamics with friction.
+    friction = tf.constant(self.friction, dtype = tf.float32)
+    # TODO: Allow option for exponentially decaying learning rate, or similar.
```
Comment (dwadden) [inline edward/inferences/sghmc.py]: For SGLD, we found that using a constant learning rate performed better than using exponential decay. Perhaps this should be an option passed into the inference method?

File: edward/inferences/sghmc.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    """
+    Utility function to calculate model's log joint density,
+    log p(x, z), for inputs z (and fixed data x).
+
+    Parameters
+    ----------
+    z_sample : dict
+      Latent variable keys to samples.
+    """
+    # TODO: This appears to be identical across HMC, SGHMC, SGLD. Should it be
```
Comment (dwadden) [inline edward/inferences/sghmc.py]: The `_log_joint` function looks like it's the same. Refactor?

File: examples/normal_sghmc.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+import edward as ed
+import tensorflow as tf
+import numpy as np
+from matplotlib import pyplot as plt
+plt.style.use("ggplot")
+
+from edward.models import Empirical, MultivariateNormalFull
+
+# Helper functions
+# TODO: Where should these go?
```
Comment (dwadden) [inline examples/normal_sghmc.py]: I added some helper functions to visualize the contours of a bivariate Gaussian (or more generally, any Edward distribution with a pdf method). Should this be refactored?

File: examples/normal_sghmc.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    X, Y = np.meshgrid(xs, ys)
+    T = tf.convert_to_tensor(np.c_[X.flatten(), Y.flatten()], dtype=tf.float32)
+    Z = sess.run(tf.exp(z.log_prob(T))).reshape((len(xs), len(ys)))
+    if ax is None:
+        fig, ax = plt.subplots()
+    cs = ax.contour(X, Y, Z)
+    if label:
+        plt.clabel(cs, inline=1, fontsize=10)
+
+def retrieve_samples(qz):
```
Comment (dwadden) [inline examples/normal_sghmc.py]: It took a bit of work to figure out how to grab the actual MCMC samples. Is there a method to do this that we missed, or should we move this elsewhere to allow for more general use?

Comment (dustinvtran) [inline examples/normal_sghmc.py]: if you just want the samples you can do `qz.params`

File: edward/inferences/sghmc.py (hunk 2 above)
Comment (dustinvtran) [inline edward/inferences/sghmc.py]: sure, that makes sense. note one difficulty i had when implementing SGLD was that i could not leverage any of the existing tensorflow optimizers, and if more learning rates were an option, one would need to implement them all from scratch. if there were a way to use the tensorflow optimizers, it would be nice so long as the implementation doesn't ruin the readability and/or speed of the code.

File: edward/inferences/sghmc.py (hunk 1 above)
Comment (dustinvtran) [inline edward/inferences/sghmc.py]: i'm not sure i understood the question but for example:
```python
>>> x = tf.constant(0)
>>> y = x
>>> x, y
(<tf.Tensor 'Const:0' shape=() dtype=int32>, <tf.Tensor 'Const:0' shape=() dtype=int32>)
```
y has a reference to the same tensor and doesn't recreate it from scratch, if that's what you're asking.

File: edward/inferences/sghmc.py (hunk 3 above)
Comment (dustinvtran) [inline edward/inferences/sghmc.py]: yeah, that sounds useful to refactor (although i'm not sure where the best location is to put it)

File: examples/normal_sghmc.py (hunk 4 above)
Comment (dustinvtran) [inline examples/normal_sghmc.py]: that's awesome. maybe we can leave it here for now and think about refactoring it in another PR that focuses more on general visualizations for edward

File: edward/inferences/sghmc.py (hunk 2 above)
Comment (dwadden) [inline edward/inferences/sghmc.py]: Got it re: tensorflow.

I'll make a new PR that adds an option to keep the learning rate constant, along with an example showing cases where this may be a bit easier to work with.

File: edward/inferences/sghmc.py (hunk 3 above)
Comment (dwadden) [inline edward/inferences/sghmc.py]: OK. Will think about this and submit another PR.

File: examples/normal_sghmc.py (hunk 4 above)
Comment (dwadden) [inline examples/normal_sghmc.py]: Cool.

File: examples/normal_sghmc.py (hunk 5 above)
Comment (dwadden) [inline examples/normal_sghmc.py]: Nice, thanks

**Comment by dwadden** [comment]
Did a bit of code cleanup - fixed indentation to 2 spaces in `normal_sghmc.py` and got rid of the TODO's since we've now addressed how to handle them.
Made a minor mistake and squashed those commits into the previous one so that the original changes we discussed don't show up. Now I know not to do that.
Let me know if there's anything else you'd like changed. Otherwise I'll start on the PR for adaptive step sizes.

File: edward/inferences/sghmc.py (hunk 1 above)
Comment (dwadden) [inline edward/inferences/sghmc.py]: Yep, that was the question. Just wanted to make sure it wasn't recreating.

**Comment by dwadden** [comment]
Cleaned things up a bit. Got rid of the README notes since we've now discussed them. Squashed into previous commit, which I now know is a bad thing to do because the changes we discussed are gone. Lesson learned.

Let me know if there are other things you'd like changed on this PR. Otherwise, I'll start thinking about the other thing we discussed: step size, code viz, and refactor of _log_joint.

**Comment by dustinvtran** [comment]
looks good to me. can you add an example using data? e.g., linear/logistic regression is fine.

**Comment by dwadden** [comment]
Will do.

**Comment by dwadden** [comment]
Git style question: I'm want to update this branch with the latest changes from master, which contains relevant bug fixes for #412 . I'm also going to add my integer type fix. 

In general, this can be done by merging master into this branch, or rebasing the branch onto master. The problem with the former approach is it will make the history messy, while doing the latter makes it impossible to view changes that were made before the rebase (like what's happened above).

Which do you prefer?

**Comment by dustinvtran** [comment]
i personally prefer to rebase to keep the commit history clean. that said, we're always squashing+merging pull requests, so feel free to do either.

**Comment by dwadden** [comment]
Cool, rebase works for me.

**Comment by dustinvtran** [comment]
i ran the examples in tensorflow 0.11.0 with python 2 and 3. i also double-checked upon merging that they work in tensorflow 1.0.0rc0. merging now.
